src/core/histogram.lua

```diff
@@ -0,0 +1,189 @@
+-- histogram.lua -- a histogram with logarithmic buckets
+--
+-- API:
+--   histogram.new(min, max) => histogram
+--     Make a new histogram, with buckets covering the range from MIN to MAX.
+--     The range between MIN and MAX will be divided logarithmically.
+--
+--   histogram.create(name, min, max) => histogram
+--     Create a histogram as in new(), but also map it into
+--     /var/run/snabb/PID/NAME, exposing it for analysis by other processes.
+--     If the file exists already, it will be cleared.
+--
+--   histogram.open(pid, name) => histogram
+--     Open a histogram mapped as /var/run/snabb/PID/NAME.
+--
+--   histogram.add(histogram, measurement)
+--     Add a measurement to a histogram.
+--
+--   histogram.iterate(histogram, prev)
+--     When used as "for bucket, lo, hi, count in histogram:iterate()",
+--     visits all buckets in a histogram.  BUCKET is the bucket index,
+--     starting from zero, LO and HI are the lower and upper bounds of
+--     the bucket, and COUNT is the number of samples recorded in that
+--     bucket.  Note that COUNT is an unsigned 64-bit integer; to get it
+--     as a Lua number, use tonumber().
+--
+--     If PREV is given, it should be a snapshot of the previous version
+--     of the histogram.  In that case, the COUNT values will be
+--     returned as a diffference between their values in HISTOGRAM and
+--     their values in PREV.
+--
+--   histogram.snapshot(a, b)
+--     Copy out the contents of A into B and return B.  If B is not given,
+--     the result will be a fresh histogram.
+--
+--   histogram.clear(a)
+--     Clear the counters in A.
+--
+--   histogram.wrap_thunk(histogram, thunk, now)
+--     Return a closure that wraps THUNK, but which measures the difference
+--     between calls to NOW before and after the thunk, recording that
+--     difference into HISTOGRAM.
+--
+module(...,package.seeall)
+
+local ffi = require("ffi")
+local shm = require("core.shm")
+local log, floor, max, min = math.log, math.floor, math.max, math.min
+
+-- Fill a 4096-byte page with buckets.  4096/8 = 512, minus the three
+-- header words means 509 buckets.  The first and last buckets are catch-alls.
+local histogram_t = ffi.typeof([[struct {
+   double minimum;
+   double growth_factor_log;
+   uint64_t count;
+   uint64_t buckets[509];
+}]])
+
+local function compute_growth_factor_log(minimum, maximum)
+   assert(minimum > 0)
+   assert(maximum > minimum)
+   -- 507 buckets for precise steps within minimum and maximum, 2 for
+   -- the catch-alls.
+   return log(maximum / minimum) / 507
```

Comment:
Any totally compelling reason not to make 509 a named constant? I realize non-4096-byte-pages aren't a huge priority, but I like to avoid exposed magic numbers when feasible. (And things like 507 = 509 - 2 are basic constant folding.)